keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+Third-party components that are both cluster admin and application developer facing can depend on
+this feature, use it, and build on top of it.
+
+### Use a Node Object Instead of Introducing a New NodeMaintenance API
+
+As an alternative, it would be possible to signal the node maintenance by marking the node object
+instead of introducing a new API. But, it is probably better to decouple this from the node for
+reasons of extensibility and complexity.
+
+Advantages of the NodeMaintenance API approach:
+- It allows us to implement incremental scale down of pods by various attributes according to a
+  drainPlan across multiple nodes.
+- There may be additional features that can be added to the NodeMaintenance in the future.
+- It helps to decouple RBAC permissions and general update responsibility from the node object.
+- It is easier to manage a NodeMaintenance lifecycle compared to the node object.
+- Two or more different actors may want to maintain the same node in two different overlapping time
+  slots. Creating two different NodeMaintenance objects would help with tracking each maintenance
+  along with the reason behind it.
+- Observability is better achieved with an additional object.
+
+### Use Taint Based Eviction for Node Maintenance
+
+To signal the start of the eviction we could simply taint a node with the `NoExecute` taint. This
+taint should be easily recognizable and have a standard name, such as
+`node.kubernetes.io/maintenance`. Other actors could observe the creations of such a taint and
+migrate or delete the pod. To ensure pods are not removed prematurely, application owners would
+have to set a toleration on their pods for this maintenance taint. Such applications could also set
+`.spec.tolerations[].tolerationSeconds`, which would give a deadline for the pods to be removed by
+the NoExecuteTaintManager.
+
+This approach has the following disadvantages:
+- Taints and tolerations do not support PDBs, which is the main mechanism for preventing voluntary
+  disruptions. People who want to avoid the disruptions caused by the maintenance taint would have
+  to specify the toleration in the pod definition and ensure it is present at all times. This would
+  also have an impact on the controllers, who would have to pollute the pod definitions with these
+  tolerations, even though the users did not specify them in their pod template. The controllers
+  could override users' tolerations, which the users might not be happy about. It is also hard to
+  make such behaviors consistent across all the controllers.
+- Taints are used as a mechanism for involuntary disruption; to get pods out of the node for some
+  reason (e.g. node is not ready). Modifying the taint mechanism to be less harmful
+  (e.g. by adding a PDB support) is not possible due to the original requirements.
+- It is not possible to incrementally scale down according to pod priorities, labels, etc.
```

Comment:
I don't see an alternative discussed in [SIG Apps meeting from 4/29](https://youtu.be/AKnVLLAKqWA) during which Clayton proposed extending PDBs with the missing functionality like:
- backpreassure
- supporting single-replica apps
- others? (I've asked for a list of PDBs problems above).

Reply:
These alternatives are discussed in the Evacuation KEP: https://github.com/atiratree/kube-enhancements/blob/improve-node-maintenance/keps/sig-apps/4212-declarative-node-maintenance/README.md#alternatives

This KEP mentions all the issues, but the implementation part is more high-level in this one.

Reply:
I have scheduled a discussion for the next sig-apps call to discuss these options.